Mini clean up!

avaidyam opened the pull request:
Removed 3 stub classes/files, which weren't used at all. (`TUIStyledView`, `TUIView+NSTextInputClient`, and `TUIKit`).

I also removed the need to set `AtleastLion`, because using `Gestalt` only in one class (`TUIScrollView`) should more than suffice as a replacement.

avaidyam committed:
Removed TUIStyledView.

It was a fairly useless class, and could be replicated using view.layer
calls to manipulate the CALayer directly.

avaidyam committed:
Removed stub category on TUIView.

avaidyam committed:
Removed unneccesary header items, and the need to set AtleastLion.

jwilling commented on .gitignore:
```diff
@@ -4,3 +4,5 @@ xcuserdata
 build
 *.perspectivev3
+
+TwUITests/specta
```
Why did you do this?

avaidyam replied:
For some reason, github.app refused to commit anything after I upgraded the specta xcodeproj. :\ I'll reverse it using the website now.

avaidyam commented:
Would it still be within the scope of TwUI to add a `CGContextDrawConicalGradient()` function to draw axial/conical gradients (a feature missing in many UI engines), because I'll be needing it for `TUISlider`?

avaidyam committed:
Removed deprecated method.

avaidyam committed:
Removed AtleastLion from example code.

jspahrsummers commented on TwUITests/specta:
```diff
@@ -1 +0,0 @@
-Subproject commit 948bfa115a0a422f1d254797e49bcc1468f03606
```
You should not have removed this submodule.

avaidyam replied:
I have no idea how to re-insert it. Would you please explain? :[

jspahrsummers commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -56,6 +57,17 @@ @implementation TUIScrollView
+	Gestalt(gestaltSystemVersionMajor, &major);
+	Gestalt(gestaltSystemVersionMinor, &minor);
+	
+	isAtleastLion = (major == 10 && minor >= 7) ? YES : NO;
```
This should account for the case of a major version greater than 10.

avaidyam replied:
Well, I'd hope Mac OS 11 would be more revolutionary than that, but sure. :P

jspahrsummers commented:
Added some notes.

@galaxas0 Please open an issue to discuss unrelated things.

avaidyam committed:
Fixed Gestalt for isAtleastLion.

avaidyam commented:
Fixed all of that! :D

jspahrsummers commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -56,6 +57,17 @@ @implementation TUIScrollView
 @synthesize resizeKnobSize;
 
+static BOOL isAtleastLion = YES;
++ (void)initialize {
```
This method also needs to check the class of `self`, since subclasses may invoke `+[TUIScrollView initialize]` if they don't override it:

``` objc
+ (void)initialize {
    if (self.class != TUIScrollView.class) return;

    // initialization code goes here
}
```

jspahrsummers commented:
Sorry, one more note. This could also use a merge from master.

avaidyam committed:
Added reentrancy guard to isAtleastLion

This avoids many subclass +initialize calls resetting isAtleastLion
unneccesarily.

avaidyam commented:
Added a reentrancy guard using `dispatch_once()` so subclasses can't call
this static variable setting over and over. Also merged upstream.

jspahrsummers commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -56,6 +57,26 @@ @implementation TUIScrollView
+	// apply a dispatch_once reentrancy guard. After all,
+	// this only needs to be set once.
+	static dispatch_once_t onceToken;
+	dispatch_once(&onceToken, ^{
```
This isn't a conventional pattern for `+initialize`. Please see my previous comment and make this match.

avaidyam replied:
But that would mean that even a `TUITableView` wouldn't initialize this
`isAtleastLion`. Would you mind explaining the logic? I'm having a hard
time grasping why I'd need to do this.

jspahrsummers replied:
`TUITableView` **shouldn't** initialize it, because `TUIScrollView` already does. It's a static variable – exactly one copy of it exists in the compiled binary.

avaidyam replied:
Then wouldn't it be better to move this to `+load`, since that's only ever
_ever_ called once?

jspahrsummers replied:
`+initialize` is nice because it's invoked lazily, so there's no impact if the class is never used.

This shouldn't use `dispatch_once` – it should use a class check (as I mentioned originally).

jspahrsummers commented:
In addition to the above comment, this is still out of date. See the pull request status on github.com.

jwilling commented:
@galaxas0 

![Tabs](http://i.imgur.com/bkKcE.png)

avaidyam commented:
Jesus Christ, @jwilling, THANK YOU SO MUCH. What the heck has my brain been
doing.

avaidyam commented:
This should work perfectly now. `+load` is only ever called once per class when the dynamic library is loaded.

jspahrsummers commented:
@galaxas0 Please update your code to use `+initialize`, and in the manner I previously described.

avaidyam commented:
Yep, merged upstream and reverted to your behavior.

avaidyam commented:
@jspahrsummers What was the issue with this again?

jspahrsummers commented:
@galaxas0 Needs to be updated with master.

jspahrsummers commented:
Looks like some unrelated code also made it into this branch, significantly expanding its scope. This should be reduced back to just the code removals and `AtLeastLion` fix.

avaidyam commented:
That would be from the merging upstream. I somehow managed to bork up git and it "copied" commits over, rather than making one merge commit.

avaidyam commented:
Merged upstream.

jspahrsummers commented:
Still out of date. See the pull request status on github.com.

avaidyam committed:
Merge upstream.

avaidyam commented:
@jwilling @jspahrsummers I've fixed the upstream merge.

avaidyam committed:
Added selective constraint removal. This was supposed to be there already, but I dun goofed.

avaidyam committed:
Fixed CA_COLOR_OVERLAY_DEBUG. Silly if statement.

jspahrsummers commented:
:mailbox:

avaidyam committed:
Thought I fixed this, but I guess not.

avaidyam commented:
Alright, @jspahrsummers I've followed your advice! :]

jspahrsummers commented:
Thanks!
